### PR TITLE
Fix (U)IntPtr construction

### PR DESCRIPTION
**Pull request by filmor**
### What does this implement/fix? Explain your changes.

Allows creating an `IntPtr` from Python. This is currently impossible, because we rely on automatic conversion (as `IntPtr` is a primitive type) while we don't actually implement one for this type (rightly so!).

The relevant code section is:
https://github.com/pythonnet/pythonnet/blob/87d4db92f6da00b4be27b842ac2ba9875569e04d/src/runtime/Types/ClassObject.cs#L73-L89

### Does this close any currently open issues?

#1116 

### Any other comments?

...

### Checklist

Check all those that are applicable and complete.

-   [ ] Make sure to include one or more tests for your change
-   [ ] If an enhancement PR, please create docs and at best an example
-   [ ] Ensure you have signed the [.NET Foundation CLA](https://cla.dotnetfoundation.org/pythonnet/pythonnet)
-   [ ] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [ ] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

### CODE REVIEW

**Comment by filmor** [comment]
I'm not entirely happy with this, yet. I think we might want to do the special-casing in `ClassObject` instead, like it's done for enums and bigints. Thoughts?

**Comment by lostmsu** [comment]
Re: IntPtr

Why can't we just treat IntPtr as non-primitive type? Then its constructor will be invoked by normal means.

**Comment by lostmsu** [comment]
Maybe we should have an explicit set of types for which this constructor masking behavior should be used, instead of relying on `IsPrimitive` which was created for a totally different purpose.

**Comment by filmor** [comment]
> Re: IntPtr
> 
> Why can't we just treat IntPtr as non-primitive type? Then its constructor will be invoked by normal means.

Because then it's always default constructed, apparently calling a cobstructor on a primitive type after it has been created "uninitialised" is a no-op. I can check again, though I'm quite certain that that was among the first things I've tried.

**Comment by lostmsu** [comment]
> apparently calling a cobstructor on a primitive type after it has been created "uninitialised" is a no-op

@filmor this would be highly problematic depending on what types this behavior apply to. If it is just `IsPrimitive` types, then we just need to adjust handling explicitly. However, if it is the case for any value types, we are in trouble.

File: tests/test_conversion.py [hunk 1]
```diff
@@ -679,3 +679,20 @@ def test_iconvertible_conversion():
...
+    assert ob.UIntPtrField == UIntPtr.Zero
+
+    ob.IntPtrField = IntPtr(Int64(-1))
+    assert ob.IntPtrField.ToInt64() == -1
+
+    ob.IntPtrField = IntPtr(Int64(1024))
+    assert ob.IntPtrField.ToInt64() == 1024
+
+    ob.UIntPtrField = UIntPtr(UInt64(1024))
+    assert ob.UIntPtrField.ToUInt64() == 1024
```
Comment (lostmsu) [inline tests/test_conversion.py:698]: It should also be possible to construct `UIntPtr` and `IntPtr` from Python `int` type with overflow exception.

**Comment by lostmsu** [comment]
@filmor this PR was rushed. There are few artifacts of development left in the final code, the latest commits never reviewed, and the earliest not reviewed completely.

File: src/runtime/Types/ClassObject.cs [hunk 2]
```diff
@@ -112,6 +104,127 @@ static NewReference tp_new_impl(BorrowedReference tp, BorrowedReference args, Bo
...
+                    }
+                    else
+                    {
+                        Exceptions.SetError(Exceptions.OverflowError, "value not in range for UIntPtr");
+                        return default;
+                    }
+                }
+            }
+
+            if (result == null && !Converter.ToManaged(op, type, out result, true))
```
Comment (lostmsu) [inline src/runtime/Types/ClassObject.cs:220]: I think we need a test for `IntPtr` constructor call with this path (e.g. parameter is neither CLR object nor Python `int`

File: src/runtime/Types/ClassObject.cs [hunk 3]
```diff
@@ -112,6 +104,127 @@ static NewReference tp_new_impl(BorrowedReference tp, BorrowedReference args, Bo
...
+        /// All numeric types and Boolean can be handled by a simple
+        /// conversion, (U)IntPtr has to be handled separately as we
+        /// do not want to convert them automically to/from integers.
+        /// </summary>
+        /// <param name="type">.NET type to construct</param>
+        /// <param name="tp">Corresponding Python type</param>
+        /// <param name="args">Constructor arguments</param>
+        private static NewReference NewPrimitive(BorrowedReference tp, BorrowedReference args, Type type)
+        {
+            // TODO: Handle IntPtr
```
Comment (lostmsu) [inline src/runtime/Types/ClassObject.cs:146]: dead TODO?

Comment (filmor) [reply src/runtime/Types/ClassObject.cs:146]: True, I'll remove it.

**Comment by filmor** [comment]
> @filmor this PR was rushed. There are few artifacts of development left in the final code, the latest commits never reviewed, and the earliest not reviewed completely. This would be OK if there was no active response for a few days, but simply bypassing review on new code is not a good idea.

It might have been a bit rushed, but that was mostly because I wanted to tackle the other remaining issues (regressions from the constructor refactor) and get a new rc out soon for people to test with. This one would have actually bitten myself for a change if we had released this as stable. :)

**Pull request merged**